### PR TITLE
Loop control

### PR DESCRIPTION
**Pull request by stepajin**
1) Added gifAnimationDuration property on GIFAnimatable, accessible when prepareForAnimation is done - added completion callback to prepareForAnimation method
2) Added loop count control

### CODE REVIEW

File: Source/Classes/GIFAnimatable.swift [hunk 1]
```diff
@@ -47,33 +52,33 @@ extension GIFAnimatable {
   /// Prepare for animation and start animating immediately.
   ///
   /// - parameter imageName: The file name of the GIF in the main bundle.
```
Comment (kaishin) [inline Source/Classes/GIFAnimatable.swift]: Documentation needs to be updated here.

File: Source/Classes/GIFAnimatable.swift [hunk 2]
```diff
@@ -47,33 +52,33 @@ extension GIFAnimatable {
...
-  public func animate(withGIFNamed imageName: String) {
-    animator?.animate(withGIFNamed: imageName, size: frame.size, contentMode: contentMode)
+  public func animate(withGIFNamed imageName: String, loopCount: Int = 0) {
+        animator?.animate(withGIFNamed: imageName, size: frame.size, contentMode: contentMode, loopCount: loopCount)
   }
 
   /// Prepare for animation and start animating immediately.
   ///
   /// - parameter imageData: GIF image data.
```
Comment (kaishin) [inline Source/Classes/GIFAnimatable.swift]: Documentation needs to be updated here.

File: Source/Classes/GIFAnimatable.swift [hunk 3]
```diff
@@ -47,33 +52,33 @@ extension GIFAnimatable {
...
-  public func animate(withGIFData imageData: Data) {
-    animator?.animate(withGIFData: imageData, size: frame.size, contentMode: contentMode)
+  public func animate(withGIFData imageData: Data, loopCount: Int = 0) {
+    animator?.animate(withGIFData: imageData, size: frame.size, contentMode: contentMode, loopCount: loopCount)
   }
 
   /// Prepares the animator instance for animation.
   ///
   /// - parameter imageName: The file name of the GIF in the main bundle.
```
Comment (kaishin) [inline Source/Classes/GIFAnimatable.swift]: Documentation needs to be updated here.

File: Source/Classes/GIFAnimatable.swift [hunk 4]
```diff
@@ -47,33 +52,33 @@ extension GIFAnimatable {
...
-  public func prepareForAnimation(withGIFNamed imageName: String) {
-    animator?.prepareForAnimation(withGIFNamed: imageName, size: frame.size, contentMode: contentMode)
+  public func prepareForAnimation(withGIFNamed imageName: String, loopCount: Int = 0, completionHandler: ((Void) -> Void)? = .none) {
+    animator?.prepareForAnimation(withGIFNamed: imageName, size: frame.size, contentMode: contentMode, loopCount: loopCount, completionHandler: completionHandler)
   }
 
   /// Prepare for animation and start animating immediately.
   ///
   /// - parameter imageData: GIF image data.
```
Comment (kaishin) [inline Source/Classes/GIFAnimatable.swift]: Documentation needs to be updated here.

File: Source/Classes/FrameStore.swift [hunk 5]
```diff
@@ -177,6 +190,11 @@ private extension FrameStore {
   /// Increments the `currentFrameIndex` property.
   func incrementCurrentFrameIndex() {
     currentFrameIndex = increment(frameIndex: currentFrameIndex)
+    if currentFrameIndex == frameCount - 1 && currentLoop == loopCount - 1 {
```
Comment (kaishin) [inline Source/Classes/FrameStore.swift]: I'd extract these conditions into 2 computed properties `isLastFrame` and `isLastLoop`

File: Source/Classes/FrameStore.swift [hunk 6]
```diff
@@ -3,6 +3,18 @@ import ImageIO
...
+  /// Total duration of one animation loop
+  var loopDuration: TimeInterval = 0
+    
+  /// Flag indicating if number of loops has been reached
+  var isFinished: Bool = false
+    
+  /// Desired number of loops, <= 0 for infinite loop
+  let loopCount: Int
+    
+  /// Number of current loop
```
Comment (kaishin) [inline Source/Classes/FrameStore.swift]: This probably should be `Index of current loop`

File: Source/Classes/Animator.swift [hunk 7]
```diff
@@ -115,8 +125,8 @@ public class Animator {
   /// - parameter imageData: GIF image data.
   /// - parameter size: The target size of the individual frames.
   /// - parameter contentMode: The view content mode to use for the individual frames.
```
Comment (kaishin) [inline Source/Classes/Animator.swift]: Documentation needs to be updated here.

File: Source/Classes/Animator.swift [hunk 8]
```diff
@@ -105,8 +115,8 @@ public class Animator {
   /// - parameter imageName: The file name of the GIF in the main bundle.
   /// - parameter size: The target size of the individual frames.
   /// - parameter contentMode: The view content mode to use for the individual frames.
```
Comment (kaishin) [inline Source/Classes/Animator.swift]: Documentation needs to be updated here.

File: Source/Classes/Animator.swift [hunk 9]
```diff
@@ -56,23 +66,23 @@ public class Animator {
...
 
-    prepareForAnimation(withGIFData: data, size: size, contentMode: contentMode)
+    prepareForAnimation(withGIFData: data, size: size, contentMode: contentMode, loopCount: loopCount, completionHandler: completionHandler)
   }
 
   /// Prepares the animator instance for animation.
   ///
   /// - parameter imageData: GIF image data.
   /// - parameter size: The target size of the individual frames.
   /// - parameter contentMode: The view content mode to use for the individual frames.
```
Comment (kaishin) [inline Source/Classes/Animator.swift]: Documentation needs to be updated here.

File: Source/Classes/Animator.swift [hunk 10]
```diff
@@ -56,23 +66,23 @@ public class Animator {
   /// - parameter imageName: The file name of the GIF in the main bundle.
   /// - parameter size: The target size of the individual frames.
   /// - parameter contentMode: The view content mode to use for the individual frames.
```
Comment (kaishin) [inline Source/Classes/Animator.swift]: Documentation needs to be updated here.

File: Source/Classes/FrameStore.swift (hunk 5 above)
Comment (kaishin) [inline Source/Classes/FrameStore.swift]: Could be functions that take a frame index or a loop index.

**Comment by kaishin** [comment]
@stepajin Thanks for the changes!

**Comment by stepajin** [comment]
Sure, thanks for your work too. Simple yet powerful framework, easy to contribute.

**Pull request merged**